src/app/pages/PerpetualPage/components/TradingChart/streaming.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * TradingChart "realtime" data stream.
+ *
+ * Implementation of TradingView Charting Library (v18.043) subscribeBars() and unsubscribeBars():
+ * https://github.com/tradingview/charting_library/wiki/JS-Api/f62fddae9ad1923b9f4c97dbbde1e62ff437b924#subscribebarssymbolinfo-resolution-onrealtimecallback-subscriberuid-onresetcacheneededcallback
+ *
+ * If the version of the library is updated, then modifications may
+ * be necessary to this file and the datafeed.ts file in
+ * this directory. Refer to:
+ * https://github.com/tradingview/charting_library/wiki/Breaking-Changes
+ */
+
+import { weiTo2 } from 'utils/blockchain/math-helpers';
+import { Bar } from './helpers';
+import {
+  subscription as bscSubscription,
+  decodeLogs,
+} from '../../utils/bscWebsocket';
+import { symbolMap } from './helpers';
+
+// const WebSocket = require('ws');
+// const url = 'ws://localhost:8080';
+
+// TODO: Change subscription ID to perpID + candleDuration
+
+// TODO: move to config
+const address = '0xB59bdf071508B8D1f4Bb76f18CAB01eA96E1Fa4E'.toLowerCase();
+const tradeTopic =
+  '0x07e349dbc09d1f39e3df0bcb7c5cac53d0020e2da13badb5cd6fa4450801a7f7';
```

Comment:
Can you please tell me what these two addresses do?

Reply:
The address is for the Perpetual Manager contract - I'll change it to import form utils/blockchain...
The tradeTopic is the topic hash of the trade event. I think there might be a web3 util function to generate this from the abi which would be nicer, I'll take a look.

Also, the way the log decoder is currently written, it will only decode Trade events, so I need to refactor it a bit so that it can handle other types of events if needed